File: src/algorithms/dynamic_programming/mr_k_marsh.cpp
#include <vector>
#include <iostream>

using namespace std;


/*
    [ jerry@pantagruel hackerrank ] $ clang++ -O3 -msse2 -o mr_k_marsh src/algorithms/dynamic_programming/mr_k_marsh.cpp

    [ jerry@pantagruel hackerrank ] $ time ./mr_k_marsh < data/algorithms/dynamic_programming/input/mr_k_marsh_01.txt

    14

    real    0m0.003s
    user    0m0.004s
    sys     0m0.000s

*/


int perimeter(int M, int N, vector< vector<char> > &A) {

    vector< vector<int> > B(M + 1, vector<int>(N + 1, 0));
    vector< vector<int> > C(M + 1, vector<int>(N + 1, 0));

    for(int i = 0; i < M; i++)
        B[i][N] = N;

    for(int i = 0; i < N; i++)
        C[M][i] = M;


    for(int i = M - 1; i >= 0; i--) {

        for(int j = N - 1; j >= 0; j--) {

            B[i][j] = (A[i][j] == 'x') ? j : B[i][j + 1];
            C[i][j] = (A[i][j] == 'x') ? i : C[i + 1][j];
        }
    }


    int R = 0;
    for(int i = 0; i < M; i++)
        for(int j = 0; j < N; j++)
            if (A[i][j] == '.')
                for (int k = j + 1; k < B[i][j]; k++)
                    for (int l = i + 1; l < min(C[i][j], C[i][k]); l++)
                        if (B[l][j] > k)
                            R = max(R, 2 * (l - i) + 2 * (k - j));

    return R;
}


int main() {

    int M;
    int N;
    cin >> M >> N;

    vector< vector<char> > A(M, vector<char>(N));
    for(int i = 0; i < M; i++)
        for(int j = 0; j < N; j++)
            cin >> A[i][j];


    int R = perimeter(M, N, A);
    if (R > 0)
        cout << R << endl;
    else
        cout << "impossible" << endl;
}

File: src/algorithms/dynamic_programming/mr_k_marsh.py
import os
import sys
sys.path.append(os.path.join(os.path.dirname(__file__), '../../../tools'))

import files


'''
    submitted code:

    def perimeter(M, N, A):
        B = [[0 for _ in range(N + 1)] for _ in range(M + 1)]
        C = [[0 for _ in range(N + 1)] for _ in range(M + 1)]

        for i in range(M):
            B[i][N] = N

        for i in range(N):
            C[M][i] = M

        for i in range(M - 1, -1, -1):
            for j in range(N - 1, -1, -1):
                B[i][j] = j if A[i][j] == 'x' else B[i][j + 1]
                C[i][j] = i if A[i][j] == 'x' else C[i + 1][j]

        R = 0
        for i in range(M):
            for j in range(N):
                if A[i][j] == '.':
                    for k in range(j + 1, B[i][j]):
                        for l in range(i + 1, min(C[i][j], C[i][k])):
                            if B[l][j] > k:
                                R = max(R, 2 * (l - i) + 2 * (k - j))

        return R


    def main():
        M, N = [int(i) for i in input().split()]
        A    = [input() for _ in range(M)]
        R    = perimeter(M, N, A)
        print(R if R > 0 else 'impossible')


    if __name__ == "__main__":
        main()

'''

def perimeter(M, N, A):
    B = [[0 for _ in range(N + 1)] for _ in range(M + 1)]
    C = [[0 for _ in range(N + 1)] for _ in range(M + 1)]

    for i in range(M):
        B[i][N] = N

    for i in range(N):
        C[M][i] = M

    for i in range(M - 1, -1, -1):
        for j in range(N - 1, -1, -1):
            B[i][j] = j if A[i][j] == 'x' else B[i][j + 1]
            C[i][j] = i if A[i][j] == 'x' else C[i + 1][j]

    R = 0
    for i in range(M):
        for j in range(N):
            if A[i][j] == '.':
                for k in range(j + 1, B[i][j]):
                    for l in range(i + 1, min(C[i][j], C[i][k])):
                        if B[l][j] > k:
                            R = max(R, 2 * (l - i) + 2 * (k - j))

    return R


def main(argv):
    lines = files.read_lines(argv[0])
    M, N  = [int(i) for i in lines[0].split()]
    A     = lines[1:]
    R     = perimeter(M, N, A)
    print(R if R > 0 else 'impossible')


if __name__ == "__main__":
    main(sys.argv[1:])
